src/state-access.h

```diff
@@ -464,8 +483,26 @@ class state_access : public i_state_access<state_access, pma_entry> {
         aliased_aligned_write(hpage + hoffset, val);
     }
 
-    void do_write_memory(uint64_t paddr, const unsigned char *data, uint64_t log2_size) {
-        m_m.write_memory(paddr, data, UINT64_C(1) << log2_size);
+    bool do_read_memory(uint64_t paddr, unsigned char *data, uint64_t length) const {
```

Comment:
Can we assume this will work? Can this call ever be issued with an invalid data/length combination?

Reply:
I think before VirtIO this would never fail, because no device was performing memory writes on addresses choosing by guest drivers, but now this is a thing with VirtIO. 

VirtIO guest drivers will pass physical memory addresses for VirtIO devices buffers, and if the driver is mess up these addresses, this must return false so the driver can recover up and reset.

So yes, this function can be issued with invalid data/length combinations from VirtIO devices.

Reply:
So I guess we either replicate the PMA search logic there, or we deal with exceptions...